README.md
# stores

[![NPM Version][npm-badge]][npm-url] [![Build Status][travis-image]][travis-url] [![Dependency Status][gemnasium-image]][gemnasium-url] [![Gittip][gittip-image]][gittip-url]

<br>

<p>
  <img width="690" height="224" src="https://github.com/ngryman/stores/raw/master/stores.jpg" alt="stores">
  <br>
  <sup>Made with <a href="https://github.com/nosir/obelisk.js">obelisk.js</a>.</sup>
</p>

<br>

**stores** is a fast, reliable, smart caching system heavily based on streams.

**stores** deals transparently with [cache stampede] or *cache miss storm* for you.
Basically, it ensures that only **one request fetches the resource to be cached**.
Other ones are put in a queue and served ASAP, when the resource has been fetched or cached.

**stores** aims to be a solid caching facility that allows you to easily implement your custom backing store.
That said, it comes with a balanced `filesystem` store and an [express] middleware.

[cache stampede]: http://en.wikipedia.org/wiki/Cache_stampede
[express]: http://expressjs.com

## Installation

```bash
npm install stores --save --production
```

## Usage

You're in hurry, I can understand that.<br>
Here is an example on how you can cache processed images easily:

```javascript
app.get(/\.(?:jpg|png|gif)$/, stores('file', function(req, slot, next) {
	processImage(req).stream().pipe(slot);
});
```

This code ensures that you only process a given image **once**, and serve the cached version to all others.<br>
It implicitly use the `FileStore` to cache the image to the filesystem.<br>
`slot` represents a stream pointing to a cache *bucket* that the `FileStore` has automatically created.

### `stores(store, [options], [fetch])`

This provided middleware allows you to configure a store globally.

#### `store`

This mandatory argument can be:
 - a `string` that defines which type of store you want to use (for now only `file`).
 - an existing `Store` instance.
 - a `constructor` that inherits from `Store`.

```javascript
stores('file');
stores(new FileStore());
stores(FileStore);
```

Note that you can register [custom stores] and then use the `string` flavor:
```javascript
stores.MyPreciousStore = MyPreciousStore;
stores('myPrecious');
```

#### `options`

Optional, `options` will configure the store. All stores accepts those values:
 - `maxPending`: *(default: 100)*, defines how many requests can be enqueued while a resource is being cached (see [hit hot]).
 - `writesRetries`: *(default: 3)*, defines how many times the store will try to write to a cache bucket that fails (see [sealed buckets].

```javascript
stores('file', { maxPending: Infinity, writesRetries: 0 });
```

Notes that each implementation can add additional values for their specific needs (see [file store options]).

#### `fetch(req, slot, next)`

Optional, this callback allows the store to fetch missing data and store it. It is called on a cache miss, with the following arguments:
 - `req`: the network request.
 - `slot`: a `Writable` stream that will hold cache data. It is allocated by the store. You must `pipe` to it.
 - `next`: next middleware in the stack. It's useful if something went wrong and you want to abort the request asap.

```javascript
stores('file', function(req, slot, next) {
	var stream = createSomeStream(req);
	stream.pipe(slot);
});
```

Note that if you don't specify a `fetch` callback here, then you must specify it in some other way.

### `Store` object

As seen previously, you can instanciate a store by yourself for some additional flexibility. As a user, you only are interested in the public `api`.

#### `Store([options])`

Creates a store with the given options.
```javascript
var store = new FileStore({ writesRetries: 0 });
```

#### `Store#fetch(fetch)`

This tells the store how to fetch a missing resource from the cache. If you haven't specified it via the `stores` middleware, you still can do it here globally.

```javascript
var store = new FileStore();
store.fetch(function(req, slot, next) {
	var stream = createSomeStream(req);
	stream.pipe(slot);
});
```

The main advantage of specifying it this way, is that you can change the `fetch` method whenever you want during the lifecycle of your application.

#### `Store#get(req, res, next, [fetch])`

This is the most important method. It will try to fetch a resource from the cache or call `fetch` callback.<br>
It can be useful if for some reason, you prefer using `stores` in one of your existing route of middleware, or when you want to add additional logic before using it.

```javascript
var store = new FileStore();

app.get(/\.(?:jpg|png|gif)$/, function(req, res, next) {
	// don't process/cache tuhmbnails
	if (~req.url.indexOf('thumb')) {
		fs.createReadStream(path.join(root, req.url)).pipe(res);
		return;
	}

	// process others
	store.get(req, res, next, function(req, slot, next) {
		processImage(req).stream().pipe(slot);
	});
});
```

Note that if you already have specified a fetch method globally, you can then omit the `fetch` argument:
```javascript
app.get(/\.(?:jpg|png|gif)$/, function(req, res, next) {
	store.get(req, res, next);
});
```

[custom stores]: #custom-stores

## Graceful filesystem streams

You also need be aware that `stores` uses [graceful-fs-stream] (`gfs`) as dependency. `gfs` slightly changes the behavior of `fs.createReadStream` and `fs.createWriteStream` by opening / creating the underlying file on **first read or write**. The main advantage is that instead of throwing an error, those function will emit an `error` event instead.

If you do want to still use standard versions, you can still use `fs._createReadStream` or `fs._createWriteStream`.

## Custom stores

To implement a custom store, you have to inherit from the `Store` object.

This object provides two methods, `_get` and `_lock` that are respectively needed to fetch the original resource or to lock a new cache *bucket*. A cache bucket can be seen as the physical location where your cached resource will be
stored. It can be a file, a memory chunk, a REDIS key, a S3 bucket, or whatever you want.

### `Store` object

## Author

| [![twitter/ngryman](http://gravatar.com/avatar/2e1c2b5e153872e9fb021a6e4e376ead?size=70)](http://twitter.com/ngryman "Follow @ngryman on Twitter") |
|---|
| [Nicolas Gryman](http://ngryman.sh) |

[npm-badge]: http://img.shields.io/npm/v/stores.svg
[npm-url]: https://www.npmjs.org/package/stores
[travis-image]: http://img.shields.io/travis/ngryman/stores.svg
[travis-url]: https://travis-ci.org/ngryman/stores
[gemnasium-image]: http://img.shields.io/gemnasium/ngryman/stores.png
[gemnasium-url]: https://gemnasium.com/ngryman/stores
[gittip-image]: http://img.shields.io/gittip/ngryman.svg
[gittip-url]: https://www.gittip.com/ngryman


CHANGELOG.md
# Changelog

## v0.0.2

- `FileStore`, a balanced filesystem cache implementation.
- Refactored code for consistency.

## v0.0.1

- first implementation of `Store` facility.


test/index.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

require('./stores');
require('./store/store');
require('./store/file');

test/store.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

/* jshint expr: true */
'use strict';

/**
 * Module dependencies.
 */
var stores = require('..'),
	Store = stores.Store,
	Bucket = stores.Bucket,
	Readable = require('stream').Readable,
	Writable = require('stream').Writable,
	SlowStream = require('slow-stream'),
	ServerResponse = require('http').ServerResponse,
	request = require('supertest'),
	express = require('express'),
	curry = require('curry'),
	util = require('util'),
	fs = require('graceful-fs-stream'),
	path = require('path'),
	chai = require('chai'),
	sinon = require('sinon'),
	should = chai.should();

chai.use(require('sinon-chai'));

/**
 * Test constants.
 */

var fixturesPath = path.join(__dirname, 'fixtures'),
	cachePath = path.join(fixturesPath, 'cache'),
	proseText = fs.readFileSync(path.join(fixturesPath, 'prose.html'), 'utf8');

/**
 * Test helpers.
 */

/**
 * Basic implementation of a file system store for tests.
 * It can be overridden almost everywhere. That allow tweaking some spots for testing purpose.
 *
 * @param {object} [options] Options.
 * @constructor
 */
function TestStore(options) {
	options = options || {};

	Store.call(this, options);

	this._get = options.get || function(bucket, callback) {
		bucket.filename = path.join(cachePath, bucket.id + '.html');
		fs.exists(bucket.filename, function(exists) {
				var slot;
				if (exists)
					slot = fs.createReadStream(bucket.filename);
				callback(null, slot);
			});
	};

	this._lock = options.lock || function(bucket, callback) {
		setImmediate(function() {
			if (options.lockDelay)
				return setTimeout(lock, 100);
			lock();
		});

		function lock() {
			var slot = fs.createWriteStream(bucket.filename);
			callback(null, slot);
		}
	};

	// event emitter throw errors by default, silence him!
	this.on('error', function() {});
}

util.inherits(TestStore, Store);

function createApp(middleware) {
	var app = express();

	// useful to track request
	var id = 0;
	app.use(function(req, res, next) {
		req.id = id++;
		next();
	});

	if (middleware instanceof Store) {
		app.use(function(req, res, next) {
			var fetch = !middleware._fetch ? fetchTest : null;
			middleware.get(req, res, next, fetch);
		});
	}
	else
		app.use(middleware);

	// avoids logging error
	app.use(function(err, req, res, next) {
		res.send(err.status);
	});

	return app;
}

function getMiddleware(url, middleware) {
	var app = createApp(middleware);
	return request(app).get(url);
}

function getStore(url, fetch, store) {
	return getMiddleware(url, stores(store || TestStore, undefined !== fetch ? fetch : fetchTest));
}

function fetchTest(req, slot) {
	fs.createReadStream(fixturesPath + req.url).pipe(slot);
}

var fire = curry(function(app, url, text, count, cache, done) {
	cache = cache.split(', ');

	for (var i = 0; i < count; i++) {
		var test = request(app).get(url).expect('X-Cache', cache[0]);
		if (cache[1])
			test = test.expect('X-Cache-Hit', cache[1]);
		test.expect(text, done);
	}
});

var sync = curry(function(calls, done, err) {
	if (err) return done(err);
	sync.calls = sync.calls || 0;
	sync.calls++;
	if (calls == sync.calls) {
		sync.calls = 0;
		done();
	}
});

var rmdir = curry(function(dir, done) {
	var list = fs.readdirSync(dir);
	for (var i = 0; i < list.length; i++) {
		var filename = path.join(dir, list[i]);
		var stat = fs.statSync(filename);

		if (stat.isDirectory()) {
			// rmdir recursively
			rmdir(filename);
		}
		else {
			// rm filename
			fs.unlinkSync(filename);
		}
	}
	done();
});

/**
 * Test suite.
 */

describe('store', function() {

	before(rmdir(cachePath));

	afterEach(rmdir(cachePath));

	it('should call fetch when cache miss', function(done) {
		var called = false;

		getStore('/prose.html', function(req, slot, next) {
			req.should.have.property('url', '/prose.html');
			slot.should.be.instanceOf(Readable);
			next.should.be.a('function');
			fetchTest(req, slot, next);
			called = true;
		}).expect(proseText, function(err, res) {
			if (err) return done(err);
			called.should.be.true;
			done(err);
		});
	});

	it('should call inline fetch when available instead of global fetch', function(done) {
		var store = new TestStore(),
			spy = sinon.spy();

		store.fetch(spy);

		getMiddleware('/prose.html', function(req, res, next) {
			store.get(req, res, next, fetchTest);
		}).expect(proseText, function(err, res) {
			if (err) return done(err);
			spy.should.not.have.been.called;
			done();
		});
	});

	it('should cache hit a second request', function(done) {
		getStore('/prose.html')
			.expect('X-Cache', 'miss')
			.expect(proseText, function(err, res) {
				if (err) return done(err);
				getStore('/prose.html')
					.expect('X-Cache', 'hit')
					.expect(proseText, done);
			});
	});

	it('should lock a resource that is being fetched', function(done) {
		var store = new TestStore(),
			fetchCalls = 0;

		store.fetch(function() {
			fetchCalls++;
			fetchTest.apply(null, arguments);
		});

		var app = createApp(store);
		var test = fire(app, '/prose.html', proseText);
		var sync6 = sync(6, function(err) {
			if (err) return done(err);
			fetchCalls.should.equal(1);
			done();
		});

		test(1, 'miss', sync6);
		setImmediate(test.bind(null, 5, 'hit', sync6));
	});

	it('should send hot fetched data to all pending request when cache is ready', function(done) {
		var store = new TestStore({ lockDelay: true });

		var app = createApp(store);
		var test = fire(app, '/prose.html', proseText);
		var sync6 = sync(6, done);

		test(1, 'miss', sync6);
		setTimeout(test.bind(null, 5, 'hit, hot', sync6), 0);
	});

	it('should send cached data to all remaining pending requests when cache was ready', function(done) {
		var store = new TestStore();

		store.fetch(function(req, slot, next) {
			fs.createReadStream(fixturesPath + req.url)
				.pipe(new SlowStream({ maxWriteInterval: 100 }))
				.pipe(slot);
		});

		var app = createApp(store);
		var test = fire(app, '/prose.html', proseText);
		var sync6 = sync(6, done);

		test(1, 'miss', sync6);
		setTimeout(test.bind(null, 5, 'hit, deferred', sync6), 100);
	});

	describe('errors', function() {

		it('should send a 404 when cache is not accessible', function(done) {
			var store = new TestStore({ get: function(bucket, callback) {
				callback(new Error('enter the void'));
			}});

			getMiddleware('/prose.html', store)
				.expect(404)
				.expect('Not Found', done);
		});

		it('should send a 500 when reading cache fail', function(done) {
			var store = new TestStore({ get: function(bucket, callback) {
				var slot = fs.createReadStream('void');
				callback(null, slot);
			}});

			getMiddleware('/prose.html', store)
				.expect(500)
				.expect('Internal Server Error', done);
		});

		it('should send a 404 when no fetch is specified', function(done) {
			getStore('/prose.html', null)
				.expect(404)
				.end(function(err, res) {
					if (err) return done(err);
					fs.readdirSync(cachePath).should.have.lengthOf(0);
					done();
				});
		});

		it('should send a 500 when locking a cache bucket fail', function(done) {
			var store = new TestStore({ lock: function(bucket, callback) {
				callback(new Error('enter the void'));
			}});

			getMiddleware('/prose.html', store)
				.expect(500)
				.expect('Internal Server Error', done);
		});

		it('should send a 404 when cache write fails writeRetries times', function(done) {
			var store = new TestStore({ writeRetries: 1, lock: function(bucket, callback) {
				var slot = fs.createWriteStream('/i/do/not/exist');
				callback(null, slot);
			}});

			var app = createApp(store);
			var test = fire(app, '/prose.html');

			test(proseText, 1, 'miss', function(err) {
				if (err) return done(err);
				test(proseText, 1, 'miss', function(err) {
					if (err) return done(err);
					test('Not Found', 1, 'miss', function(err, res) {
						if (err) return done(err);
						res.status.should.equal(404);
						done();
					});
				});
			});
		});

	});

	describe('events', function() {

		it('should emit a miss event', function(done) {
			var store = new TestStore(), called = false;
			store.on('miss', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				called = true;
			});

			getMiddleware('/prose.html', store).expect(proseText, function(err, res) {
				if (err) return done(err);
				called.should.be.true;
				done();
			});
		});

		it('should emit a hit event', function(done) {
			var store = new TestStore({ lockDelay: true }),
				classic = 0,
				hot = 0,
				deferred = 0;

			store.fetch(function(req, slot, next) {
				fs.createReadStream(fixturesPath + req.url)
					.pipe(new SlowStream({ maxWriteInterval: 100 }))
					.pipe(slot);
			});

			store.on('hit', function(req, bucket, type) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);

				switch (type) {
					case 'hot': hot++; break;
					case 'deferred': deferred++; break;
					default: classic++; break;
				}
			});

			var app = createApp(store);
			var test = fire(app, '/prose.html', proseText);
			var sync4 = sync(4, function(err) {
				if (err) return done(err);
				classic.should.equal(1);
				hot.should.equal(1);
				deferred.should.equal(1);
				done();
			});

			test(1, 'miss', sync4);
			setTimeout(test.bind(null, 1, 'hit, hot', sync4), 0);
			setTimeout(test.bind(null, 1, 'hit, deferred', sync4), 150);
			setTimeout(test.bind(null, 1, 'hit', sync4), 400);
		});

		it('should emit bucket\'s locked/pending/lock/unlock events', function(done) {
			var store = new TestStore(),
				locked = 0,
				pending = 0,
				lock = 0,
				unlock = 0;

			store.on('bucket:locked', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				locked++;
			}).on('bucket:pending', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				pending++;
			}).on('bucket:lock', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				lock++;
			}).on('bucket:unlock', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				unlock++;
			});

			var app = createApp(store);
			var test = fire(app, '/prose.html', proseText);
			var sync3 = sync(3, function(err) {
				// ensures every event is fired
				setTimeout(function() {
					if (err) return done(err);
					locked.should.equal(2);
					pending.should.equal(2);
					lock.should.equal(1);
					unlock.should.equal(1);
					done();
				}, 50);
			});

			test(1, 'miss', sync3);
			test(2, 'hit', sync3);
		});

		it('should emit bucket\'s seal/sealed events', function(done) {
			var store = new TestStore({ writeRetries: 1, lock: function(bucket, callback) {
				var slot = fs.createWriteStream('/i/do/not/exist');
				callback(null, slot);
			}}),
				seal = 0,
				sealed = 0;

			store.on('bucket:seal', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				seal++;
			}).on('bucket:sealed', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				sealed++;
			});

			var app = createApp(store);
			var test = fire(app, '/prose.html');

			test(proseText, 1, 'miss', function(err) {
				if (err) return done(err);
				test(proseText, 1, 'miss', function(err) {
					if (err) return done(err);
					test('Not Found', 1, 'miss', function(err) {
						if (err) return done(err);
						seal.should.equal(1);
						sealed.should.equal(1);
						done();
					});
				});
			});
		});

		it('should emit bucket\'s new/get/release/destroy events', function(done) {
			var store = new TestStore({ lockDelay: true }),
				newB = 0,
				getB = 0,
				release = 0,
				destroy = 0;

			store.on('bucket:new', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				newB++;
			}).on('bucket:get', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				getB++;
			}).on('bucket:release', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				release++;
			}).on('bucket:destroy', function(req, bucket) {
				req.should.exist;
				bucket.should.be.instanceOf(Bucket);
				destroy++;
			});

			var app = createApp(store);
			var test = fire(app, '/prose.html', proseText);
			var sync2 = sync(2, function(err) {
				if (err) return done(err);
				newB.should.equal(1);
				getB.should.equal(2);
				release.should.equal(3);
				destroy.should.equal(1);
				done();
			});

			test(1, 'miss', function(err) {
				if (err) return done(err);
				test(2, 'hit', sync2);
			});
		});

	});

});

test/stores.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

/* jshint expr: true */
'use strict';

/**
 * Module dependencies.
 */
var stores = require('..'),
	chai = require('chai'),
	sinon = require('sinon'),
	should = chai.should();

chai.use(require('sinon-chai'));

describe('stores middleware', function() {

	it('should accept a Store type', function() {
		var Store = sinon.spy();
		stores(Store);
		Store.should.have.been.calledWithNew;
	});

	it('should accept a Store instance', function() {
		var store = new stores.Store();
		stores(store);
	});

	it('should accept the name of a Store type', function() {
		stores.WootStore = sinon.spy();
		stores('woot');
		stores.WootStore.should.have.been.called;
		delete stores.WootStore;
	});

	it('should accept an optional fetch function', function() {
		var Store = sinon.spy();
		Store.prototype.fetch = sinon.spy();
		var fetch = function() {};
		stores(Store, fetch);
		Store.should.have.been.calledWithNew;
		Store.prototype.fetch.should.have.been.calledWith(fetch);
	});

	it('should accept options', function() {
		var Store = sinon.spy(), options = { woot: 'store' };
		stores(Store, options);
		Store.should.have.been.calledWith(options);
	});

	it('should accept all arguments', function() {
		var Store = sinon.spy();
		Store.prototype.fetch = sinon.spy();
		var fetch = function() {};
		var options = { woot: 'store' };
		stores(Store, options, fetch);
		Store.should.have.been.calledWithNew;
		Store.should.have.been.calledWith(options);
		Store.prototype.fetch.should.have.been.calledWith(fetch);
	});

});

lib/lock.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

var queues = {};

/**
 *
 * @type {{}}
 */
var lock = {};

/**
 *
 * @param bucket
 * @returns {boolean}
 */
lock.locked = function(bucket) {
	return (null != queues[bucket.id]);
};

/**
 *
 * @param bucket
 */
lock.lock = function(bucket) {
	queues[bucket.id] = [];
};

/**
 *
 * @param bucket
 */
lock.unlock = function(bucket) {
	queues[bucket.id] = null;
};

/**
 *
 * @param bucket
 * @param req
 * @param res
 * @param next
 */
lock.pending = function(bucket, req, res, next) {
	bucket._req = req;
	bucket._res = res;
	bucket._next = next;
	queues[bucket.id].push(bucket);
};

/**
 *
 * @param bucket
 * @param fn
 */
lock.clear = function(bucket, fn) {
	queues[bucket.id].forEach(fn);
	queues[bucket.id].length = 0;
};

/**
 * Exports.
 */
module.exports = lock;

Gruntfile.js
/*!
 * stores
 * Copyright (c) 2013 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

module.exports = function(grunt) {
	// loads npm tasks
	require('load-grunt-tasks')(grunt);
	// time tasks
	require('time-grunt')(grunt);

	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		meta: {
			banner: '/*! <%%= pkg.name %> - v<%%= pkg.version %> - ' +
				'<%%= grunt.template.today("yyyy-mm-dd") %>\n' +
				'<%%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
				'* Copyright (c) <%%= grunt.template.today("yyyy") %> <%%= pkg.author.name %>;' +
				' Licensed <%%= pkg.license %> */\n'
		},
		env: {
			test: {
				NODE_ENV: 'development'
			}
		},
		jshint: {
			options: grunt.file.readJSON('.jshintrc'),
			all: ['lib/*.js', 'test/*.js', 'Gruntfile.js']
		},
		mochaTest: {
			options: {
				reporter: 'spec',
				bail: true,
				slow: 2000
			},
			all: ['test/index.js']
		}
	});

	// front tasks
	grunt.registerTask('test', ['env:test', 'jshint', 'mochaTest']);
};

lib/bucket.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */
var crypto = require('crypto');

function Bucket(id, store) {
	this.id = id;
	this.store = store;
	this.container = store._buckets;
	this.refCount = 1;
	this.writeErrors = 0;
}

Bucket.prototype.locked = function(req) {
	var locked = (null != this.container[this.id]._queue);
	if (locked)
		this.store.emit('bucket:locked', req, this);
	return locked;
};

Bucket.prototype.lock = function(req) {
	this.container[this.id]._queue = [];
	this.store.emit('bucket:lock', req, this);
};

Bucket.prototype.unlock = function(req) {
	this.container[this.id]._queue = null;
	this.store.emit('bucket:unlock', req, this);
	this.release(req);
};

Bucket.prototype.pending = function(req) {
	this.container[this.id]._queue.push(req);
	this.store.emit('bucket:pending', req, this);
};

Bucket.prototype.clearPending = function(fn) {
	var bucket = this.container[this.id];
	bucket._queue.forEach(fn);
	bucket._queue.length = 0;
};

Bucket.prototype.release = function(req) {
	this.refCount--;
	this.store.emit('bucket:release', req, this);
	if (0 === this.refCount && 0 === this.writeErrors) {
		this.container[this.id] = null;
		this.store.emit('bucket:destroy', req, this);
	}
};

Bucket.prototype.sealed = function(req) {
	var sealed = this.container[this.id]._sealed;
	if (sealed)
		this.store.emit('bucket:sealed', req, this);
	return sealed;
};

Bucket.prototype.seal = function(req) {
	this.container[this.id]._sealed = true;
	this.store.emit('bucket:seal', req, this);
};

Bucket.get = function(req, store) {
	var id = crypto.createHash('sha256')
		.update(req.url)
		.digest('hex');

	if (!store._buckets)
		store._buckets = {};

	var container = store._buckets,
		bucket = container[id];

	if (bucket) {
		bucket.refCount++;
		store.emit('bucket:get', req, bucket);
	}
	else {
		bucket = new Bucket(id, store);
		store.emit('bucket:new', req, bucket);
		container[id] = bucket;
	}

	return bucket;
};

/**
 * Exports.
 */

module.exports.Bucket = Bucket;

lib/stores.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

/**
 * Middleware.
 *
 * @param {Store|string} store
 * @param {function} [fetch]
 * @param {object} [options]
 * @returns {Function}
 */

function stores(store, options, fetch) {
	// arguments juggling
	if ('function' == typeof options) {
		fetch = options;
		options = null;
	}

	// if string, fetches flavor + Store constructor
	if ('string' == typeof store) {
		var name = store[0].toUpperCase() + store.slice(1) + 'Store';
		store = stores[name];
	}

	// if constructor, instantiates it
	if ('function' == typeof store)
		/* jshint newcap: false */
		store = new store(options);

	// global fetch
	if (fetch)
		store.fetch(fetch);

	// middleware!
	return function(req, res, next) {
		store.get(req, res, next);
	};
}

/**
 * Exports.
 */

stores.Bucket = require('./bucket').Bucket;
stores.Store = require('./store/store').Store;
stores.FileStore = require('./store/file').FileStore;

module.exports = stores;

lib/store/file.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

// TODO: handle http methods: https://github.com/koajs/static-cache/blob/master/index.js

/**
 * Module dependencies.
 */

var Store = require('./store').Store,
	fs = require('graceful-fs-stream'),
	path = require('path'),
	util = require('util'),
	mkdirp = require('mkdirp');

/**
 *
 * @param {object} [options]
 * @constructor
 */
function FileStore(options) {
	Store.call(this, options);

	this.root = options.root || process.cwd();
	this.depth = options.depth || 4;
}

util.inherits(FileStore, Store);

/**
 * Fetches cache data if it exists.
 *
 * @param {Bucket} bucket
 * @param {function} callback
 * @private
 */
FileStore.prototype._get = function(bucket, callback) {
	var filename = path.resolve(
		path.join(this.root, buildFilename(bucket.id, this.depth))
	);

	// if file exists, passes a readable stream
	fs.exists(filename, function(exists) {
		var slot;
		if (exists)
			slot = fs.createReadStream(filename);
		callback(null, slot);
	});

	// stores file info in the bucket, for further use
	bucket.filename = filename;
	bucket.pathname = path.dirname(filename);
};

/**
 * Creates the directory structure and a stream to the file cache.
 *
 * @param {Bucket} bucket
 * @param {function} callback
 * @private
 */
FileStore.prototype._lock = function(bucket, callback) {
	// creates the directory structure, opens the new file
	// and passes a writable stream
	mkdirp(bucket.pathname, function(err) {
		if (err) return callback(err);

		var slot = fs.createWriteStream(bucket.filename);
		callback(null, slot);
	});
};

/**
 * Creates a balanced directory structure for each cache file.
 * Storing too much files in the same directory has a negative impact in
 * most of filesystems.
 * This ensure that there is not more that 256 directory per depth, expect for the final directory.
 * We use the id (a strong hash) to generate the filename.
 * So the final directory should not contain so much files neither.
 * Balancing is taken care of naturally by the hash.
 *
 * @param {string} id
 * @param {number} depth
 * @return {string}
 * @private
 */
function buildFilename(id, depth) {
	var filename = '',
		i = 0,
		length = id.length;

	for ( ; i < depth && i < length; i += 2)
		filename += id.substr(i, 2) + path.sep;
	filename += id.substr(i);

	return filename;
}

/**
 * Exports.
 */

module.exports.FileStore = FileStore;

lib/store/store.js
/*!
 * stores
 * Copyright (c) 2014 Nicolas Gryman <ngryman@gmail.com>
 * MIT Licensed
 */

'use strict';

/**
 * Module dependencies.
 */

var Bucket = require('../bucket').Bucket,
	EventEmitter = require('events').EventEmitter,
	PassThrough = require('stream').PassThrough,
	util = require('util');

var DEBUG = ('development' == process.env.NODE_ENV);

/**
 * Store facility.
 *
 * @param {object} [options]
 * @constructor
 */
function Store(options) {
	options = options || {};

	this.maxPending = options.maxPending || 100;
	this.writeRetries = options.writeRetries || 3;
}

util.inherits(Store, EventEmitter);

/**
 *
 * @param {object} req
 * @param {object} res
 * @param {function} next
 * @param {function} [fetch]
 */
Store.prototype.get = function(req, res, next, fetch) {
	var bucket = Bucket.get(req, this);

	// debug headers
	if (DEBUG)
		res.on('header', function() {
			res.header('X-Cache', req.cacheHit ? 'hit' : 'miss');
			if ('string' == typeof req.cacheHit)
				res.header('X-Cache-Hit', req.cacheHit);
		});

	// if bucket is sealed, resource is flagged as not accessible,
	// so we send a 404
	if (bucket.sealed(req))
		return onError.call(this, req, bucket, 404, new Error('cache bucket sealed'));

	// try to get a resource from the cache
	setImmediate(
		this._get.bind(this,
			bucket,
			onGet.bind(this, req, bucket, fetch)
		)
	);
};

/**
 *
 * @param {function} fetch
 * @return {Store}
 */
Store.prototype.fetch = function(fetch) {
	this._fetch = fetch;
	return this;
};

/**
 *
 * @param {object} req
 * @param {Bucket} bucket
 * @param {function} fetch
 * @param {Error} err
 * @param {Readable} slot
 * @private
 */
function onGet(req, bucket, fetch, err, slot) {
	if (err) return onError.call(this, req, bucket, 404, err);

	// already locked, set the bucket as pending
	// from now on it will wait until it's unlocked
	if (bucket.locked(req))
		return bucket.pending(req);

	/** hit */

	// pipes bucket's slot to the response, and we're done
	if (slot) {
		req.cacheHit = req.cacheHit || true;
		this.emit('hit', req, bucket, 'string' == typeof req.cacheHit ? req.cacheHit : undefined);
		slot.pipe(req.res);
		slot.on('error', onError.bind(this, req, bucket, 500));
		bucket.release(req);
		return;
	}

	/** miss */

	// first checks that we have a way to fetch fresh data
	fetch = fetch || this._fetch;
	if (!fetch)
		return onError.call(this, req, bucket, 404, new Error('no fetch method specified'));

	// associate a lock to this bucket, avoiding other request to lock and fetch it
	bucket.lock(req);

	// locks a writable cache slot.
	// Fetched data will be written to that slot.
	setImmediate(
		this._lock.bind(this,
			bucket,
			onLock.bind(this, req, bucket, fetch)
		)
	);
}

/**
 *
 * @param {object} req
 * @param {Bucket} bucket
 * @param {function} fetch
 * @param {Error} err
 * @param {Writable} slot
 * @private
 */
function onLock(req, bucket, fetch, err, slot) {
	if (err) return onError.call(this, req, bucket, 500, err);

	// emits a miss event right now
	this.emit('miss', req, bucket);

	// we'll directly broadcast fetched data...
	var broadcast = new PassThrough();
	broadcast.setMaxListeners(this.maxPending);

	// ... to the current response
	broadcast.pipe(req.res);

	// ... to the cache slot, in order to store it
	broadcast.pipe(slot);

	// when the slot is ready
	slot.on('finish', function() {
		// emits a cache even right now
		this.emit('cache', req, bucket);

		// we tell other pending buckets to get it from the store again
		bucket.clearPending(onHitDeferred.bind(this, bucket));

		// finally unlock the bucket, it's ready!
		bucket.unlock(req);
	}.bind(this));

	// when an error occurred
	slot.on('error', function() {
		// increment write errors count
		bucket.writeErrors++;

		// too many write errors, seals the bucket
		if (bucket.writeErrors > this.writeRetries)
			bucket.seal(req);

		// we tell other pending buckets to get it from the store again
		bucket.clearPending(onHitDeferred.bind(this, bucket));

		// finally unlock the bucket, it's ready!
		bucket.unlock(req);
	}.bind(this));

	// ... to all other pending requests
	// this ensure a fast response for requests that arrived
	// between bucket's lock and slot's lock
	bucket.clearPending(onHitHot.bind(this, bucket, broadcast));

	// brainz!
	fetch(req, broadcast, onError.bind(this, req, bucket, 500));
}

/**
 *
 * @param {Bucket} bucket
 * @param {Readable} slot
 * @param {object} req
 * @private
 */
function onHitHot(bucket, slot, req) {
	// emits a hit event, type hot
	this.emit('hit', req, bucket, 'hot');
	// marks as hip hop
	req.cacheHit = 'hot';
	// releases bucket as we're done
	bucket.release(req);
	// pipes slot to response
	slot.pipe(req.res);
}

/**
 *
 * @param {Bucket} bucket
 * @param {object} req
 * @private
 */
function onHitDeferred(bucket, req) {
	// note: we don't emit an event here as we call store.get

	// marks as hit deferred
	req.cacheHit = 'deferred';
	// releases bucket as we're done
	bucket.release(req);
	// get from cache
	this.get(req, req.res, req.next);
}

/**
 *
 * @param {object} req
 * @param {Bucket} bucket
 * @param {number} status
 * @param {Error} err
 */
function onError(req, bucket, status, err) {
	// sets http status
	err.status = status;
	// emits a error event
	this.emit('error', err);
	// releases bucket as we're done
	bucket.release(req);
	// passes to next error handler
	req.next(err);
}

/**
 * Exports.
 */

module.exports.Store = Store;